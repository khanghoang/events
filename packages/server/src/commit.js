/* @flow */
import { createError } from 'micro-boom';
import { json } from 'micro';
import redis from 'redis';

import type { CommitConfig } from '../types/Config.type';
import { createClient, RedisClientType } from './redis-client';
import runLua from './runLua';

type Event = {
  type: string,
  payload: mixed,
};

export const commit = async (redisClient: RedisClientType, event: Event, namespc: string) => {
  const lua = `
    local event = ARGV[1];
    local counter = redis.call('INCR', KEYS[1]);

    redis.call('HSET', KEYS[2], counter, event);
    redis.call('PUBLISH', ARGV[2]..'::events', counter .. ':' .. event);
    return counter;
  `;

  return runLua(redisClient, lua, {
    keys: [`{${namespc}}::id`, `{${namespc}}::events`],
    argv: [JSON.stringify(event), namespc],
  });
};


export default ({ redis, namespc }: CommitConfig) => {
  const client = createClient(redis);

  return async (req: any) => {
    const event = await json(req);

    if (typeof event.type !== 'string') {
      throw createError(422, 'Missing type');
    }

    const id = await commit(client, event, namespc);

    return { ...event, id };
  };

}
