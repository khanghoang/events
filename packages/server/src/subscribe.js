/* @flow */
import kefir from 'kefir';
import once from 'lodash/fp/once';
import range from 'lodash/fp/range';

import type { SubscribeConfig } from '../types/Config.type';
import { createClient } from './redis-client';
import { query } from './query';

const tryParse = raw => {
  try {
    return JSON.parse(raw);
  } catch (ex) {
    return { type: '@@RAW', payload: raw };
  }
};

function flush(response) {
  if (response.flush && response.flush.name !== 'deprecated') {
    response.flush();
  }
}

const toOutput = events => `id: ${events[events.length - 1].id}
event: INCMSG
data: ${JSON.stringify(events)}

`;

export default ({ redis, history, debug, namespc, burst }: SubscribeConfig) => {
  const subClient = createClient(redis);
  const queryClient = createClient(redis);

  // prepare cache
  const MAX_SIZE = history.size;
  const list = [];
  const cache = {};

  const addToCache = message => {
    debug && console.log('adding to cache', message.id);

    const newSize = list.unshift(message.id);
    cache[message.id] = message;

    if (newSize > MAX_SIZE) {
      const expired = list.splice(MAX_SIZE);
      expired.forEach(removingId => {
        delete cache[removingId];
      });
    }

    debug && console.log('HISTORY', list);
    debug && console.log('CACHE\n', cache);
  };

  const message$ = kefir
    .fromEvents(subClient, 'message', (channel, message) => message)
    .map(message => {
      const rawId = message.split(':')[0];
      const id = Number(rawId);
      const rawEvent = message.slice(rawId.length + 1);

      return {
        id,
        ...tryParse(rawEvent),
      };
    })
    .onValue(addToCache);

  const addClient = (src, opts, match, req, res) => {
    req.socket.setTimeout(0);
    req.socket.setNoDelay(true);
    req.socket.setKeepAlive(true);

    res.writeHead(200, {
      'Content-Type': 'text/event-stream;charset=UTF-8',
      'Cache-Control': 'no-cache',
      Connection: 'keep-alive',
    });

    res.write(':ok\n\n');

    // cast to integer
    const retry = opts.retry | 0;

    if (retry) {
      res.write('retry: ' + retry + '\n');
    }

    flush(res);

    const subscription = src
      .filter(x => x)
      .bufferWithTimeOrCount(burst.time, burst.count)
      .filter(b => b.length)
      .map(toOutput)
      .observe(block => {
        debug && console.log('send to %s %s', req.url, block);
        res.write(block);
      });

    const removeClient = once(() => {
      debug && console.log('removing', req.url);
      subscription.unsubscribe();
      res.end();
    });

    req.on('end', removeClient);
    req.on('close', removeClient);
    res.on('finish', removeClient);
  };

  const getInitialValues = lastEventId => {
    if (!lastEventId) {
      return kefir.never();
    }

    // history[0] is the latest id
    const current = list[0];
    const oldestInCache = list[list.length - 1];

    debug && console.log('current', current);
    debug && console.log('oldestInCache', oldestInCache);
    if (oldestInCache <= lastEventId + 1) {
      return kefir
        .constant(range(lastEventId + 1, current + 1).map(id => cache[id]))
        .flatten();
    }

    debug && console.error('too old, getting more from redis');
    const fromCache$ = kefir.constant(
      range(oldestInCache, current + 1).map(id => cache[id])
    );

    const promise = list.length
      ? query(queryClient, namespc, lastEventId, oldestInCache - 1)
      : query(queryClient, namespc, lastEventId);

    const fromRedis$ = kefir.fromPromise(promise);

    return kefir.concat([fromRedis$, fromCache$]).flatten();
  };

  const service = async (req: any, res: any) => {
    try {
      debug && console.log('connected', req.url);
      const lastEventId = Number(req.headers['last-event-id']);

      const initialValues = getInitialValues(lastEventId);

      debug && console.log('lastEventId', lastEventId);

      // const url = req.url;
      // const query = parse(url.split('?')[1]);

      addClient(kefir.concat([initialValues, message$]), {}, {}, req, res);
    } catch (ex) {
      console.log(ex);
      throw ex;
    }
  };

  subClient.subscribe(`${namespc}::events`);

  return {
    service,
    unsubscribe: () => {
      debug && console.log('unsubscribing from redis');
      subClient.unsubscribe(`${namespc}::events`);
    },
  };
};
