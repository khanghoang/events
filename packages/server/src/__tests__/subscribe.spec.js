/* @flow */
import { get, router } from 'microrouter';
import del from 'redis-functional/del';
import micro from 'micro';

import { delay } from 'awaiting';
import listen from 'test-listen';
import subscribe from '@events/subscriber';

import { commit } from '../commit';
import { createClient } from '../redis-client';
import makeSubscribe from '../subscribe';

jest.unmock('micro');

const takeEvents = (count, stream) =>
  stream.take(count).scan((prev, next) => prev.concat(next), []).toPromise();

describe('subscribe endpoint', () => {
  it('should subscribe', async () => {
    const pubClient = createClient();
    const { service, unsubscribe } = makeSubscribe({
      namespc: 'test-sub',
      history: { size: 10 },
    });
    const server = micro(router(get('/', service)));
    const url = await listen(server);
    const { events$, abort } = subscribe(url, {
      'Burst-Count': '10',
      'Burst-Time': '10',
    });

    // 1 -> :ok
    // 2 -> first event
    // 3 -> second event
    // 4 -> abort
    const promise = takeEvents(2, events$);
    await delay(100);

    pubClient.publish('test-sub::events', '1:{"type":"first"}');
    pubClient.publish('test-sub::events', `2:{"type":"second","payload":123}`);
    await delay(100);

    abort();
    server.close();
    unsubscribe();
    const val = await promise;
    pubClient.quit();
    expect(val).toMatchSnapshot();
  });

  it('should work with 2 clients', async () => {
    const pubClient = createClient();
    const { service, unsubscribe } = makeSubscribe({
      namespc: 'test-sub',
      redis: {
        url: process.env.REDIS_URL,
      },
      history: { size: 10 },
    });
    const server = micro(router(get('/', service)));
    const url = await listen(server);

    const headers = {
      'Burst-Count': '1',
      'Burst-Time': '10',
    };

    const http1 = subscribe(url + '?client=1', headers);
    const http2 = subscribe(url + '?client=2', headers);
    await delay(100);

    const promise1 = takeEvents(4, http1.raw$);
    const promise2 = takeEvents(3, http2.raw$);

    // publish 2 events
    // client 1 will wait until the end
    // client 2 will abort early after the first event
    pubClient.publish('test-sub::events', '1:{"type":"first"}');

    await delay(10);
    http2.abort();

    await delay(10);
    pubClient.publish('test-sub::events', `2:{"type":"second"}`);

    await delay(10);
    http1.abort();

    unsubscribe();

    const val1 = await promise1;
    const val2 = await promise2;

    expect(val1.length).toBe(2);
    expect(val2.length).toBe(1);
  });

  it('should work with empty cache', async () => {
    const commitClient = createClient({ url: process.env.REDIS_URL });

    await del(commitClient, '{test-no-cache}::events');
    await del(commitClient, '{test-no-cache}::id');
    const events = [
      { type: 'test', payload: 1 },
      { type: 'test', payload: 2 },
      { type: 'test', payload: 3 },
    ];

    await Promise.all(
      events.map(evt => commit(commitClient, evt, 'test-no-cache'))
    );

    const { service, unsubscribe } = makeSubscribe({
      namespc: 'test-no-cache',
      history: { size: 10 },
      burst: { time: 10, count: 10 },
      redis: {
        url: process.env.REDIS_URL,
      },
    });

    const server = micro(router(get('/', service)));
    const url = await listen(server);
    const { events$, abort } = subscribe(url, { 'Last-Event-ID': 1 });

    const output = await takeEvents(2, events$);

    abort();
    server.close();
    unsubscribe();

    expect(output).toMatchSnapshot();
  });
});
