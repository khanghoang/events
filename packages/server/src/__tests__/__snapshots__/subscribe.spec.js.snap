// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribe endpoint should subscribe 1`] = `
Array [
  Object {
    "id": 1,
    "type": "first",
  },
  Object {
    "id": 2,
    "payload": 123,
    "type": "second",
  },
]
`;

exports[`subscribe endpoint should work with empty cache 1`] = `
Array [
  Object {
    "id": 2,
    "payload": 2,
    "type": "test",
  },
  Object {
    "id": 3,
    "payload": 3,
    "type": "test",
  },
]
`;
